0024-off-chain-runtime-logic.md

```diff
@@ -0,0 +1,263 @@
+
+1. **Remote Attestation.** The normal runtime attestation flow is initiated by
+   the host. As a result of this flow, the `node.CapabilityTEE` structure is
+   generated which includes the remote attestation quote and additional data.
+
+2. **Node Endorsement.** The host verifies the `node.CapabilityTEE` structure
+   and if deemed correct, it signs it using the node's identity key and the
+   following domain separation context:
+
+   ```
+   oasis-core/node: ROFL TEE capability
+   ```
+
+   The signature is stored in a new structure `EndorsedCapabilityTEE` which is
+   defined as follows:
+
+   ```go
+   type EndorsedCapabilityTEE struct {
+      // CapabilityTEE is the TEE capability structure to be endorsed.
+      CapabilityTEE CapabilityTEE `json:"capability_tee"`
+
+      // NodeEndorsement is the optional node endorsement signature.
+      NodeEndorsement *Signature `json:"node_endorsement,omitempty"`
+   }
+   ```
+
+3. **Updating Node-Endorsed CapabilityTEE in ROFL.** The `EndorsedCapabilityTEE`
+   is sent to ROFL to be stored and available for establishing secure EnclaveRPC
+   sessions.
+
+4. **RONL Verification.** When establishing a new session with RONL, the
+   endorsed TEE capability is presented during session establishment. RONL
+   verifies the quote, ensures the enclave identity is one of the known
+   identities set at compile-time and verifies the node endorsement against the
+   locally known node identity (both RONL and ROFL must be from the same node).
+
+   If all the checks pass, a secure EnclaveRPC session is established.
+
+This flow needs to be repeated whenever RAK changes for any reason and also
+periodically to ensure freshness (consistent with the quote policy configured
+for the runtime in the consensus layer).
+
+### Updates to the ORC Manifest
+
+The ORC manifest is extended with a field that can specify extra components
+which currently include ROFL binaries in a similar way as we already support
+regular runtime binaries (e.g. specifying the executable and SGX metadata).
+
+### Updates to the Runtime Host Protocol
+
+This proposal includes some non-breaking updates to the Runtime Host Protocol in
+order to support the ROFL component, as follows:
+
+- **Consensus Block Notification.** No updates are required to facilitate
+  notifications about consensus layer blocks as this is already handled as part
+  of the existing RHP flow. The only change is that for ROFL, these
+  notifications invoke a hook that can be implemented by the runtime.
+
+- **Runtime Transaction Submission.** A new method `HostSubmitTx` is introduced
+  which allows ROFL to submit transactions to the runtime. It works by queueing
+  the transaction in the transaction pool (local queue) for later scheduling.
+
+  ```go
+  type HostSubmitTxRequest struct {
+      // Tx is the raw transaction data.
+      Tx []byte `json:"tx"`
+  }
+  ```
+
+- **RONL-ROFL Communication.** The existing EnclaveRPC is reused to facilitate
```

Comment:
Is the idea of this to be one-way? e.g.  ROFL can access RONL (as explained below via `ronl` identifier), or is there a use-case for the other way as well? It might be useful if RONL could invoke a notification hook in ROFL?